### PR TITLE
[Fix #4889, #5287, and #5348] - SafeNavigation fixes

### LINKED ISSUE #4889: Style/SafeNavigation throws an offense and can autocorrect into a style that violates another cop (Lint/SafeNavigationChain)
Code that violates `Style/SafeNavigation`:

```ruby
add_error(:ids, :parameter_validation, "Max limit for fetching conversations by ids is 50") if ids && ids.length > 50
```

Auto-corrects to code that violates `Lint/SafeNavigationChain`:

```ruby
add_error(:ids, :parameter_validation, "Max limit for fetching conversations by ids is 50") if ids&.length > 50
```

This is especially problematic because it’s a style cop auto-correcting to code that violates a lint cop. That is, it's recommending (or auto-correcting) code that works fine to code that actually does not work.

--------

```
$ rubocop -V
0.51.0 (using Parser 2.4.0.0, running on ruby 2.3.1 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by rrosenblum**
This fixes #4889, #5287, and #5348.

* Do not register an offense for `foo && foo.bar !~ /baz/`
`=~`, `!~`, and `<=>` are not in the list of comparable methods.
* Auto-correct all methods in a method chain to use safe navigation
This is needed for safety. Example: `foo && foo.bar.baz` if this is corrected to `foo&.bar.baz` and `foo = nil`, then `baz` would be run on `nil`.
* Do not register an offense when an unsafe method is used in a method chain
This is the same reasoning as why all methods in a chain need safe navigation.

### CODE REVIEW

File: spec/rubocop/cop/style/safe_navigation_spec.rb [hunk 1]
```diff
@@ -44,16 +44,41 @@
...
 
+    it 'allows an object check before a method call that is used in ' \
+      'a non standard comparison' do
+      expect_no_offenses('foo && foo.bar !~ /baz/')
+    end
+
+    it 'allows an object check before a method chain that is used in ' \
+      'a comparison' do
+      expect_no_offenses('foo && foo.bar > 2')
+    end
```
Comment (koic) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: It seems to be the same test as L57.
https://travis-ci.org/bbatsov/rubocop/jobs/326143335#L1090-L1097

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +26,9 @@
 * [#5357](https://github.com/bbatsov/rubocop/issues/5357): Fix `Lint/InterpolationCheck` false positives on escaped interpolations. ([@pocke][])
 * [#5409](https://github.com/bbatsov/rubocop/issues/5409): Fix multiline indent for `Style/SymbolArray` and `Style/WordArray` autocorrect. ([@flyerhzm][])
 * [#5393](https://github.com/bbatsov/rubocop/issues/5393): Fix `Rails/Delegate`'s false positive with a method call with arguments. ([@pocke][])
+* [#5348](https://github.com/bbatsov/rubocop/issues/5348): Fix false positive for `Style/SafeNavigation` when safe guarding more comparison methods. ([@rrosenblum][])
+* [#4889](https://github.com/bbatsov/rubocop/issues/4889): Auto-correcting `Style/SafeNavigation` will add safe navigation to all methods in a method chain. ([@rrosenblum][])
+* [#5287](https://github.com/bbatsov/rubocop/issues/5287<Paste>): Do not register an offense in `Style/SafeNavigation` if there is an unsafe method used in a method chain. ([@rrosenblum][])
```
Comment (koic) [inline CHANGELOG.md]: Let's remove `<Paste>`.

```diff
-[#5287](https://github.com/bbatsov/rubocop/issues/5287<Paste>)
+[#5287](https://github.com/bbatsov/rubocop/issues/5287)
```

https://ci.appveyor.com/project/bbatsov/rubocop/build/2411/job/eyjgicj6f8mfyjd2#L196

File: spec/rubocop/cop/style/safe_navigation_spec.rb [hunk 3]
```diff
@@ -44,16 +44,26 @@
...
-    it 'allows method call that are used in a comparison safe guarded by ' \
+    it 'allows method call that is used in a comparison safe guarded by ' \
       'an object check' do
       expect_no_offenses('foo.bar > 2 if foo')
     end
 
-    it 'allows an object check before a method call that are used in ' \
+    it 'allows method call that is used in a non standard comparison ' \
+      'safe guarded by an object check' do
+      expect_no_offenses('foo.bar !~ /baz/ if foo')
```
Comment (bbatsov) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: I don't see tests for `=~` and the spaceship operator. And wouldn't not necessarily call them non-standard comparisons. :-) I'd just go with `regexp matching` and `spaceship operator`.

File: lib/rubocop/cop/style/safe_navigation.rb [hunk 4]
```diff
@@ -98,7 +98,9 @@ def autocorrect(node)
           lambda do |corrector|
             corrector.remove(begin_range(node, body))
             corrector.remove(end_range(node, body))
-            corrector.insert_before((method_call || body).loc.dot, '&')
+            corrector.insert_before(method_call.loc.dot, '&')
+
+            add_saf_nav_to_all_methods_in_chain(corrector, method_call, body)
```
Comment (bbatsov) [inline lib/rubocop/cop/style/safe_navigation.rb]: saf should safe I guess.

**Comment by bbatsov** [comment]
Apart from my small remarks I think you should extend the cop's documentation to explain better what's covered by it, what's not covered and why.

**Comment by rrosenblum** [comment]
I am working on updating the documentation and cleanup some functionality. Since we only convert methods that use `.` and we now check the entire method chain for unsafe methods, the specific checks for comparison methods are no longer needed.

I also think that we can likely remove the methods that `nil` implements from the list of unsafe methods since safe navigation will prevent the method call when the receiver is `nil`. Making sure all methods in a method chain use safe navigation is also important for this.

**Comment by bbatsov** [comment]
Are you going to do this in the scope of this PR or a different PR?

**Comment by rrosenblum** [comment]
Since there are a high number of issues open for this cop, I will wrap up the documentation as a part of this PR and handle any extra cleanup and changes in a separate PR.

**Comment by rrosenblum** [comment]
I have updated the documentation. Let me know what you think or if I should add more details.

**Comment by bbatsov** [comment]
👍

**Pull request merged**